opendatahub-io/opendatahub-operator
# allow setting default deployment mode for Kserve in DSC

## Description

This change is to support Raw Deployments for Kserve. JIRA: https://issues.redhat.com/browse/RHOAIENG-3132 
This PR adds a field `DefaultDeploymentMode` to the Kserve spec in the DSC. The default value is `Serverless`, which introduces no changes in behavior. However, if the `DefaultDeploymentMode` is set to `RawDeployment`, the operator updates the configmap `inferenceservice-config` in the applications NS. The updates are as follows : 
- deploy["defaultDeploymentMode"] is set to match the value passed to the DSC since they both use the same enum
- ingress["disableIngressCreation "] is set to `true`  if the deployment mode is `RawDeployments`, otherwise it is `false` 

## How Has This Been Tested?



Changes tested against dev cluster by building and deploying `quay.io/vedantm/opendatahub-operator:rawdeployment` 
Testing instructions : 
- clone the operator repo and run `make deploy -e IMG=quay.io/vedantm/opendatahub-operator:latest -e OPERATOR_NAMESPACE=opendatahub-operator-system` 
- Create a DSCInitialization as follows
```
spec:
  applicationsNamespace: opendatahub
  monitoring:
    managementState: Managed
    namespace: opendatahub
  serviceMesh:
    controlPlane:
      metricsCollection: Istio
      name: data-science-smcp
      namespace: istio-system
    managementState: Unmanaged
``` 
- Create a DSC as follows 
```
    kserve:
      defaultDeploymentMode: RawDeployment
      managementState: Managed
      serving:
        ingressGateway:
          certificate:
            type: SelfSigned
        managementState: Removed
        name: knative-serving
``` 
- Once the DSC reconcile is successful, verify values in the `inferenceservice-config` ConfigMap in the applications NS 
![image](https://github.com/opendatahub-io/opendatahub-operator/assets/20891020/2ca18c13-cd36-49f5-814a-bfe61ca5781c)
![image](https://github.com/opendatahub-io/opendatahub-operator/assets/20891020/770cf72d-4db5-4174-9f6d-d25c490b7aad)
- Change the DSC spec such that default deployment is now `Serverless`, wait for DSC to reconcile and verify the corresponding changes in the `inferenceservice-config` object 



## Merge criteria:
